### PR TITLE
[cli] check main after master

### LINKED ISSUE #8428: Running `pulumi new` against a github repo without a master branch fails
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

When running `pulumi new` against a GitHub repo that has `main` as the default branch, Pulumi fails with the following error message:

```zsh
error: couldn't find remote ref "refs/heads/master"
```

### Steps to reproduce

Run the following: `pulumi new https://github.com/pierskarsenbarg/new-pulumi`

(that's a GitHub repo without a `master` branch)

If you create a master branch (even if it's not the default branch) it works.

Expected:  To clone and create a new project using the code in the repo
Actual:  Error message occurs

### PR DESCRIPTION
**Pull request by iwahbe**
# Description

If no branch is specified, probe `master`, then probe `main` before failing. 

Fixes #8428

I would be curious if anyone has a different idea on how to download branches from both master and main without breaking changes. 

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by justinvp** [comment]
> * Allow the user to specify the branch they want to download as part of `pulumi new` using the `pulumi new https://github.com/pierskarsenbarg/new-pulumi#main` syntax

Why do we need this? It's already possible to do this via:

```
pulumi new https://github.com/pierskarsenbarg/new-pulumi/tree/main
```

> If no branch is specified, probe master, then probe main before failing.

This makes sense to me.

**Comment by iwahbe** [comment]
> Why do we need this? It's already possible to do this via:
```
pulumi new https://github.com/pierskarsenbarg/new-pulumi/tree/main
```

I didn't catch that. I removed the new version.

**Review by pgavlin — COMMENTED** [review]
Could we add a couple unit tests for the repo URL parser? LGTM otherwise.

**Comment by iwahbe** [comment]
@pgavlin Did you have a specific thing to test in mind? We already have
https://github.com/pulumi/pulumi/blob/2c25c3fbd0bf076b2949739b0f9dda8772729ddd/sdk/go/common/util/gitutil/git_test.go#L27-L74

**Comment by pgavlin** [comment]
> Did you have a specific thing to test in mind? We already have

Nope, those preexisting tests look good!

**Review by pgavlin — APPROVED** [review]

File: CHANGELOG_PENDING.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 ### Improvements
 
+- [cli] - Allow specifying branch with `#branch` in `pulumi new`.
+  [#8463](https://github.com/pulumi/pulumi/pull/8463)
```
Comment (justinvp) [inline CHANGELOG_PENDING.md:4]: Update to reflect the changes in the PR.

**Review by justinvp — APPROVED** [review]
LGTM otherwise

**Pull request merged**